firmware/src/PeO.c

```diff
@@ -0,0 +1,102 @@
+#include "PeO.h"
+#ifdef MACHINE_ON
+    #include "machine.h"
+#endif
+#include "usart.h"
+
+volatile float max_power;
+volatile float max_power_duty_cycle;
+static uint8_t done;
+static uint8_t callSweep;
+volatile uint8_t count;
+
+void perturb_and_observe(void){
+    static uint8_t d_step = PWM_D_STEP;
+
+    // Computes power input
+    control.pi[0] = control.v_panel[0] * control.i_panel[0];
+
+    //Derivate power
+    float dpi = (control.pi[0]) -(control.pi[1]);
+    float dvi = (control.v_panel[0]) -(control.v_panel[1]);
+    float dii = (control.i_panel[0]) - (control.i_panel[1]);
+    float coef_di = (dii/dvi);
+    float coef_def = ((-control.i_panel[0]/control.v_panel[0]));
+
+ 
+    if(!callSweep){
+        if(dvi != 0){
+            if(coef_di != coef_def){
+                if(coef_di > coef_def){
+                    control.D -= d_step;
+                }else{
+                    control.D +=d_step;
+                }
+            }
+            else{
+                if(dii != 0){
+                    if(dii > 0){
+                        control.D += d_step;
+                    }
+                    else{
+                        control.D -= d_step;
+                    }
+                }
+            }
+        }
+        count++;
+        if(count == 30){
+            max_power = 0;
+            callSweep = 1;
+            done = 0;
+        }
```

Comment:
I recommend to all this code be a separated function and call it perturb_and_observe and rename this function to something like control_update and then instead of having a complex code inside the if(!callSweep)
you will have something like this:

```
void control_update(void) 
{

/* Update metrics */
  if (callSweep) {
    sweep(...)
  } else {
    perturb_and_observe(...);
  }
}
```